File: 2023/clojure/p03.clj
;; # 🎄 Advent of Code 2023 - Day 3
(ns p03)

(def data-string (slurp "../input/03.txt"))

(def test-string "467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..")

(def digits {\1 1 \2 2 \3 3 \4 4 \5 5 \6 6 \7 7 \8 8 \9 9 \0 0})

(defn ->data
  "Parse the data recursively"
  [data-string]
  (loop [loc [0 0]
         board {:syms {} :nums {}}
         s data-string]
    (let [[row col] loc
          [next & rest] s]
      (cond
        ; done
        (= next nil) board
        ; on newline, increment row
        (= next \newline) (recur [(inc row) 0] board rest)
        ; on . increment col
        (= next \.) (recur [row (inc col)] board rest)
        ; if we hit a digit, consume it and store at its loc
        (digits next) (let [digits (re-find #"\d+" (apply str s))
                            n (count digits)]
                        (recur [row (+ col n)]
                               (assoc-in board [:nums loc] (parse-long digits))
                               (drop n s)))
        ; otherwise we must have a symbol
        :else (recur [row (inc col)] (assoc-in board [:syms loc] next) rest)))))

(def data (->data data-string))
(def test-data (->data test-string))

(defn neighbors
  "Compute all of the squares that neighbor a number"
  [[row col] num]
  (let [n (count (str num))]
    (into #{}
          (concat
           ; the row above
           (for [i (range -1 (inc n))] [(dec row) (+ col i)])
           ; the row below
           (for [i (range -1 (inc n))] [(inc row) (+ col i)])
           ; left and right
           [[row (dec col)] [row (+ col n)]]))))

(defn part-1 [{nums :nums syms :syms}]
  (letfn [(touching-symbol? [[loc num]]
            (some syms (neighbors loc num)))]
    (transduce 
      (comp 
        (filter touching-symbol?) 
        (map second)) 
      + nums)))

(assert (= (part-1 test-data) 4361))

(def ans1 (part-1 data))

(assert ans1 559667)

;; ## Part 2
;;
;; Now we need to identify gears, which are *s that are touching exactly two numbers.

(defn prod [xs] (reduce * xs))

(defn part-2 [data]
  (let [{nums :nums syms :syms} data]
    (letfn [(num-neighbors [star]
              (let [[row col] star
                    ; find the numbers that are within shooting distance of the star
                    possible-nums (select-keys nums (for [c (range -3 2) r (range -1 2)] [(+ row r) (+ col c)]))]
                (into []
                      (comp
                       ; only keep those that actually touch the star
                       (filter (fn [[loc num]] ((neighbors loc num) star)))
                       ; get the number itself
                       (map second))
                      possible-nums)))]
      (transduce
       (comp
        ; find the stars
        (filter (fn [x] (= (second x) \*)))
        ; grab their locs
        (map first)
        ; find the lists of their number neighbors
        (map num-neighbors)
        ; gears need to have two neighbors
        (filter #(= (count %) 2))
        ; compute the gear-ratio
        (map prod))
       + syms))))

(assert (= (part-2 test-data) 467835))

(def ans2 (part-2 data))

(assert (= ans2 86841457))

(defn -main [_]
  (println "Answer1:" ans1)
  (println "Answer2:" ans2))

File: 2023/python/p03.py
import math
from dataclasses import dataclass
from typing import NamedTuple
from collections.abc import Generator
import re

with open("../input/03.txt") as f:
    data = f.read()

test_data = """467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."""

digits = set("1234567890")


def is_symbol(char: str) -> bool:
    return (char != ".") and (char not in digits)


class Coord(NamedTuple):
    row: int
    col: int


@dataclass
class Board:
    syms: dict[Coord, str]
    nums: dict[Coord, int]


def process_data(data: str) -> Board:
    """Convert the string to a Board with syms and nums, coords are complex."""
    syms = {
        Coord(row, col): char
        for row, line in enumerate(data.splitlines())
        for col, char in enumerate(line)
        if is_symbol(char)
    }

    nums = {
        Coord(row, match.span()[0]): int(match.group())
        for row, line in enumerate(data.splitlines())
        for match in re.finditer(r"\d+", line)
    }

    return Board(syms, nums)


def neighbors(loc: Coord, num: int) -> Generator[Coord, None, None]:
    """Return the neighbors of a word."""
    n = len(str(num))
    yield from (Coord(loc.row - 1, loc.col + i) for i in range(-1, n + 1))
    yield from (Coord(loc.row + 1, loc.col + i) for i in range(-1, n + 1))
    yield from (Coord(loc.row, loc.col - 1), Coord(loc.row, loc.col + n))


def part1(data: str) -> int:
    board = process_data(data)
    nums = board.nums
    part_nums = (
        num
        for coord, num in nums.items()
        if any(y in board.syms for y in neighbors(coord, num))
    )
    return sum(part_nums)


assert part1(test_data) == 4361

ans1 = part1(data)

assert ans1 == 559_667

# Part 2
# Now we need to find the stars that are boarding two numbers.


def part2(data: str) -> int:
    """Find the gear ratios."""
    board = process_data(data)
    # start with the star symbols
    stars = (coord for coord, sym in board.syms.items() if sym == "*")
    # we need to find the ones with two numbers nearby,
    # let's speed this up a bit by searching in a window around
    # each star, let's figure out the maximum number length we need to worry about.
    max_num_length = max(map(lambda x: len(str(x)), board.nums.values()))

    def nearby_nums(star):
        """Return the nums that are next to a star."""
        for row in range(-1, 2):
            for col in range(-max_num_length, 2):
                coord = Coord(star.row + row, star.col + col)
                if num := board.nums.get(coord):
                    yield coord, num

    # replace each star with a list of the nums they touch
    num_neighbors = (
        [num for coord, num in nearby_nums(star) if star in set(neighbors(coord, num))]
        for star in stars
    )
    # Gears only touch two nums
    gears = filter(lambda x: len(x) == 2, num_neighbors)
    # their gear ratio is the product of the nums
    ratios = (math.prod(nums) for nums in gears)
    return sum(ratios)


assert part2(test_data) == 467_835

ans2 = part2(data)

assert ans2 == 86_841_457


if __name__ == "__main__":
    print("Answer1:", ans1)
    print("Answer2:", ans2)
